tools/text/line-map.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package text
+
+import (
+	"ballerina-lang-go/common/errors"
+	"fmt"
+)
+
+// LineMap represents a collection of text lines in the TextDocument.
+type LineMap interface {
+	TextLine(line int) (TextLine, error)
+	LinePositionFromPosition(position int) (LinePosition, error)
+	TextPositionFromLinePosition(linePosition LinePosition) (int, error)
+	TextLines() []string
+}
+
+type lineMapImpl struct {
+	textLines []TextLine
+	length    int
+}
+
+func NewLineMap(textLines []TextLine) LineMap {
+	return &lineMapImpl{
+		textLines: textLines,
+		length:    len(textLines),
+	}
+}
+
+func (lm lineMapImpl) TextLine(line int) (TextLine, error) {
+	if err := lm.lineRangeCheck(line); err != nil {
+		return nil, err
+	}
+	return lm.textLines[line], nil
+}
+
+func (lm lineMapImpl) LinePositionFromPosition(position int) (LinePosition, error) {
+	if err := lm.positionRangeCheck(position); err != nil {
+		return nil, err
+	}
+	textLine := lm.findLineFrom(position)
+	return LinePositionFromLineAndOffset(textLine.LineNo(), position-textLine.StartOffset()), nil
+}
+
+func (lm lineMapImpl) TextPositionFromLinePosition(linePosition LinePosition) (int, error) {
+	if err := lm.lineRangeCheck(linePosition.Line()); err != nil {
+		return 0, err
+	}
+	textLine := lm.textLines[linePosition.Line()]
+	if textLine.Length() < linePosition.Offset() {
+		return 0, errors.NewIllegalArgumentError(fmt.Sprintf("Cannot find a line with the character offset '%d'", linePosition.Offset()))
+	}
```

Comment:
Minor thing shall we return -1 instead of 0 on error since it obviously an invalid value, where as 0 is still a valid value